lib/Optimizer/GraphOptimizer.cpp

```diff
@@ -909,6 +944,41 @@ static NodeValue simplifyConcatNode(Function *F, ConcatNode *CN) {
     }
   }
 
+  /// If the inputs of concat are all reshape nodes, and the inputs of reshape
+  /// are able to be concatenated, then we can perform concat first and then reshape.
+  /// Concat(Reshape(x) * N) -> Reshape(Concat(x * N)).
```

Comment:
This optimization comes last in `simplifyConcatNode` currently. Does this order matter?

E.g. if this optimization would create new opportunities for further concat optimizations, would the optimizer detect it and make use of it?

Specifically, I was thinking about something like:

`Concat(Reshape(Slice(X)) * N) -> Reshape(Concat(Slice(x) * N))`
In theory, the result of such a transform could be used to perform another existing optimization like
` // Concat(Slice(X, 0..10), Slice(X, 10..20)) -> X`

I don't know though if `Concat(Reshape(Slice(X)) * N) -> Reshape(Concat(Slice(x) * N))` ever happens in real networks. May be it never happens or is too rare to justify the potential additional code complexity to handle it?

I'd suggest to write a new unit-test for this example and see if both optimizations kick-in or if we more code is needed to implement this? May be it just works already?

Reply:
I go over the this process in LLDB today. I found that all the newly-created nodes by the simplifyConcatNode() are appended to end of the list of F's nodes. And the optimizeConcatNodes() optimization would iterate onto them, even if they were later added to the nodes list.

It means, the newly-created Concat nodes would also be considered for optimization. So, i think for this case `Concat(Reshape(Slice(X)) * N) -> Reshape(Concat(Slice(x) * N))`, the optimization would kick in and produce `Reshape(X)` in the end.